Include/objimpl.h

```diff
@@ -309,16 +301,16 @@ extern PyGC_Head *_PyGC_generation0;
  */
 #define _PyObject_GC_UNTRACK(o) do { \
     PyGC_Head *g = _Py_AS_GC(o); \
-    assert(_PyGCHead_REFS(g) != _PyGC_REFS_UNTRACKED); \
-    _PyGCHead_SET_REFS(g, _PyGC_REFS_UNTRACKED); \
-    g->gc.gc_prev->gc.gc_next = g->gc.gc_next; \
-    g->gc.gc_next->gc.gc_prev = g->gc.gc_prev; \
+    PyGC_Head *prev = _PyGCHead_PREV(g); \
+    assert(g->gc.gc_next != NULL); \
+    prev->gc.gc_next = g->gc.gc_next; \
+    _PyGCHead_SET_PREV(g->gc.gc_next, prev); \
     g->gc.gc_next = NULL; \
+    g->gc.gc_prev &= _PyGC_PREV_MASK_FINALIZED; \
```

Comment:
Is that the same as ` _PyGCHead_SET_PREV(g, NULL)`? The latter seems clearer to me.

Reply:
Bit 1 must be kept after untracking.
But bit 2 and 3 must be cleared when untracking, because this macro may be called while GC
(i.e. weakref callback or tp_finalize).

Reply:
Why not adding a macro setting the pointer and flags at once? _PyGCHead_SET_PREV_FLAGS(g->gc.gc_next, prev, flags).

Reply:
Because it's not useful.  gc_prev is used as:

* Set prev pointer, without touching any flags. (_PyTrash_deposit_object and _PyObject_GC_TRACK)
* Set prev pointer to list header node.  Flags are not used in list header. (_PyObject_GC_TRACK)
* Clear prev pointer and flags, but keep only `_PyGC_PREV_MASK_FINALIZED`. (only here.)
* Clear gc_prev entirely. `= (uintptr_t)0`.